source/Halibut.Tests/Transport/Streams/TimeoutStreamWrapperFixture.cs

```diff
@@ -0,0 +1,369 @@
+                }
+            }
+        }
+
+        [Test]
+        public async Task WriteAsyncShouldTimeout()
+        {
+            var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                CancellationToken, 
+                onListenerRead: async _ => await Task.Delay(TimeSpan.FromDays(1), CancellationToken));
+
+            using (disposables)
+            {
+                sut.WriteTimeout = (int)TimeSpan.FromSeconds(5).TotalMilliseconds;
+                // Ensure the correct timeout is used
+                sut.ReadTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+
+                var data = new byte[655360];
+                var r = new Random();
+                r.NextBytes(data);
+
+                var stopWatch = Stopwatch.StartNew();
+                Exception? actualException = null;
+
+                // Brute force attempt to get the Write to be slow
+                while (!CancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await sut.WriteAsync(data, 0, data.Length, CancellationToken);
+                    }
+                    catch (Exception e)
+                    {
+                        actualException = e;
+                        break;
+                    }
+                }
+
+                stopWatch.Stop();
+
+                actualException.Should().NotBeNull().And.BeOfType<OperationCanceledException>();
+                actualException!.Message.Should().Be("The WriteAsync operation timed out after 00:00:05.");
+
+                stopWatch.Elapsed.Should().BeLessThan(TimeSpan.FromSeconds(10));
+            }
+        }
+
+        [Test]
+        public async Task CloseShouldPassThrough()
+        {
+            var (disposables, sut, _) = await BuildTcpClientAndTcpListener(CancellationToken);
+
+            using (disposables)
+            {
+                sut.Close();
+
+                Action read = () => sut.Read(new byte[1], 0, 1);
+                read.Should().Throw<ObjectDisposedException>("Because the stream is closed");
+            }
+        }
+
+        [Test]
+        public async Task DisposeShouldPassThrough()
+        {
+            var (disposables, sut, _) = await BuildTcpClientAndTcpListener(CancellationToken);
+
+            using (disposables)
+            {
+                sut.Dispose();
+
+                Action read = () => sut.Read(new byte[1], 0, 1);
+                read.Should().Throw<ObjectDisposedException>("Because the stream is closed");
+            }
+        }
+
+        async Task<(IDisposable Disposables, Stream SystemUnderTest, Func<string, Task> PerformServiceWriteFunc)> BuildTcpClientAndTcpListener(
+            CancellationToken cancellationToken, 
+            Func<string, Task>? onListenerRead = null)
+        {
+            Func<string, Task>? performServiceWriteFunc = null;
+            var disposableCollection = new DisposableCollection();
```

Comment:
We run a small risk of not disposing things we add to this list if an exception happens half way through this method.

I wonder if we should pass the `disposableCollection` as a parameter instead, and declare it in the calling method with 'using' as a way to guarantee disposal?

Reply:
Or another idea, could we make this a disposable class instead?

That could clean up some of the lambdas and tuples.

Maybe. Just a thought 🙂

Reply:
Nice to have, it could leak, i could do it better, it's a test, it's ok'ish